### PR TITLE
Async Crypto Endeavour

### PR DESCRIPTION
**Pull request by dignifiedquire**
Ref https://github.com/libp2p/js-libp2p-crypto/pull/10

BREAKING CHANGE

New method PeerInfo.create(cb) replaces the previous new PeerInfo() to create a fresh id

### CODE REVIEW

File: test/peer-info.spec.js [hunk 1]
```diff
@@ -6,47 +6,63 @@ const PeerId = require('peer-id')
...
+      expect(err).to.not.exist
+      expect(pi).to.exist
+      expect(pi.id).to.exist
+      done()
+    })
   })
 
   it('add multiaddr', () => {
-    const pi = new PeerInfo()
     expect(pi).to.exist
```
Comment (diasdavid) [inline test/peer-info.spec.js]: checking something to exist every time doesn't seem useful

File: test/peer-info.spec.js [hunk 2]
```diff
@@ -6,47 +6,63 @@ const PeerId = require('peer-id')
...
     const mh = Multiaddr('/ip4/127.0.0.1/tcp/5001')
     pi.multiaddr.add(mh)
     expect(pi.multiaddrs.length).to.equal(1)
   })
 
   it('add multiaddr that are buffers', () => {
-    const pi = new PeerInfo()
     expect(pi).to.exist
```
Comment (diasdavid) [inline test/peer-info.spec.js]: same

File: test/peer-info.spec.js [hunk 3]
```diff
@@ -6,47 +6,63 @@ const PeerId = require('peer-id')
...
     const mh = Multiaddr('/ip4/127.0.0.1/tcp/5001')
     pi.multiaddr.add(mh.buffer)
     expect(pi.multiaddrs[0] instanceof Multiaddr).to.equal(true)
   })
 
   it('add repeated multiaddr', () => {
-    const pi = new PeerInfo()
     expect(pi).to.exist
```
Comment (diasdavid) [inline test/peer-info.spec.js]: same

Comment (diasdavid) [inline test/peer-info.spec.js]: and it goes on

**Comment by diasdavid** [comment]
How did you solve the PCKS#1 vs PKCS#6. Can we confirm that we are manipulating and storing the keys in the same way as in go? This will be crucial for config file share, identify and secio crypto handshake.

File: test/peer-info.spec.js (hunk 1 above)
Comment (dignifiedquire) [inline test/peer-info.spec.js]: agreed

**Comment by dignifiedquire** [comment]
> How did you solve the PCKS#1 vs PKCS#6. Can we confirm that we are manipulating and storing the keys in the same way as in go? This will be crucial for config file share, identify and secio crypto handshake.

We already had tests for this (the go interop part) here and in libp2p-crypto. You can see the details of the conversion here: https://github.com/libp2p/js-libp2p-crypto/commit/eb5de1017be263a0344dd6358ec1ca17dea87d2a

File: test/peer-info.spec.js (hunk 3 above)
Comment (dignifiedquire) [inline test/peer-info.spec.js]: fixed

File: src/index.js [hunk 4]
```diff
@@ -91,3 +91,13 @@ function Peer (peerId) {
   // TODO: add features to fetch multiaddr using filters
   // look at https://github.com/whyrusleeping/js-mafmt/blob/master/src/index.js
 }
+
+Peer.create = (cb) => {
```
Comment (diasdavid) [inline src/index.js]: Ok, to simplify, let's default to use PeerInfo.create

- [ ] s/Peer/PeerInfo
- [ ] signature becomes `(peerId, callback) => `
- [ ] Use always PeerInfo.create everywhere

File: src/index.js [hunk 5]
```diff
@@ -91,3 +91,13 @@ function Peer (peerId) {
...
+  Id.create((err, key) => {
+    if (err) {
+      return cb(err)
+    }
+
+    cb(null, new Peer(key))
```
Comment (diasdavid) [inline src/index.js]: key? or Id?

**Comment by diasdavid** [comment]
- [ ] This PR is missing the documentation update

File: src/index.js (hunk 4 above)
Comment (dignifiedquire) [inline src/index.js]: You mean with `peerId` being optional?

Comment (dignifiedquire) [inline src/index.js]: done

File: .travis.yml [hunk 6]
```diff
@@ -1,9 +1,16 @@
...
+
+matrix:
+  include:
+  - node_js: 4
+    env: CXX=g++-4.8
+  - node_js: 6
+    env: CXX=g++-4.8
+  - node_js: stable
+    env:
+    - SAUCE=true
```
Comment (diasdavid) [inline .travis.yml]: Please do Sauce on LTS

**Comment by dignifiedquire** [comment]
@diasdavid should be ready now

**Pull request merged**